### PR TITLE
Close #135 - Voltron Monitoring Service refactoring

### LINKED ISSUE #135: Voltron Monitoring Service refactoring
Depends on / implements: Terracotta-OSS/terracotta-apis#140

Needed for: #127 and ehcache/ehcache3#1008

### PR DESCRIPTION
**Pull request by mathieucarbou**
Close #135 - Voltron Monitoring Service refactoring

@anthonydahanne : needs tc-platform release __5.0.6.beta2__

### CODE REVIEW

File: management/cluster-topology/src/main/java/org/terracotta/management/model/cluster/Server.java [hunk 1]
```diff
@@ -319,4 +319,33 @@ public static Server create(String serverName) {
...
+  //  Currently the possible transitions for a server are:
+  //  UNINITIALIZED->SYNCHRONIZING->PASSIVE->ACTIVE
+  //  UNINITIALIZED->ACTIVE
+
+    UNINITIALIZED,
+    SYNCHRONIZING,
+    PASSIVE,
+    ACTIVE,
+
+    UNKNOWN,;
```
Comment (anthonydahanne) [inline management/cluster-topology/src/main/java/org/terracotta/management/model/cluster/Server.java]: no need for the colon "," here

File: management/tms-entity/tms-entity-tests/src/test/java/org/terracotta/management/entity/tms/TmsAgentTest.java [hunk 2]
```diff
@@ -105,9 +103,7 @@ public void setUp() throws Exception {
     );
     client.addConnection(connection);
 
-    connection.fetchServerEntity(expectedCluster.getStripe("stripe-1").get()
-        .getServerByName("server-1").get()
-        .getServerEntity(getClass().getSimpleName() + ":" + TmsAgentConfig.ENTITY_TYPE).get());
+    assertTrue(connection.fetchServerEntity(getClass().getSimpleName(), TmsAgentConfig.ENTITY_TYPE));
```
Comment (anthonydahanne) [inline management/tms-entity/tms-entity-tests/src/test/java/org/terracotta/management/entity/tms/TmsAgentTest.java]: I guess we have here the opportunity to test hasFetchedServerEntity; what about this : 

    assertFalse(connection.hasFetchedServerEntity(getClass().getSimpleName(), TmsAgentConfig.ENTITY_TYPE));
    assertTrue(connection.fetchServerEntity(getClass().getSimpleName(), TmsAgentConfig.ENTITY_TYPE));
    assertTrue(connection.hasFetchedServerEntity(getClass().getSimpleName(), TmsAgentConfig.ENTITY_TYPE));

File: management/tms-entity/tms-entity-tests/src/test/java/org/terracotta/management/entity/tms/TmsAgentTest.java [hunk 3]
```diff
@@ -105,9 +103,7 @@ public void setUp() throws Exception {
...
   }
 
   @After
```
Comment (anthonydahanne) [inline management/tms-entity/tms-entity-tests/src/test/java/org/terracotta/management/entity/tms/TmsAgentTest.java]: We could also test unfetchServerEntity here : 

    assertTrue(connection.hasFetchedServerEntity(getClass().getSimpleName(), TmsAgentConfig.ENTITY_TYPE));
    connection.unfetchServerEntity(getClass().getSimpleName(), TmsAgentConfig.ENTITY_TYPE);
    assertFalse(connection.hasFetchedServerEntity(getClass().getSimpleName(), TmsAgentConfig.ENTITY_TYPE));

File: management/monitoring-service/src/main/java/org/terracotta/management/service/monitoring/DefaultPlatformNotification.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * @author Mathieu Carbou
+ */
+class DefaultPlatformNotification implements PlatformNotification {
+
+  private static final long serialVersionUID = 1;
+
+  private static final AtomicLong MUTATION_INDEX = new AtomicLong(Long.MIN_VALUE);
```
Comment (anthonydahanne) [inline management/monitoring-service/src/main/java/org/terracotta/management/service/monitoring/DefaultPlatformNotification.java]: NOTIFICATION_INDEX ?

File: management/tms-entity/tms-entity-tests/src/test/java/org/terracotta/management/entity/tms/TmsAgentTest.java (hunk 2 above)
Comment (mathieucarbou) [inline management/tms-entity/tms-entity-tests/src/test/java/org/terracotta/management/entity/tms/TmsAgentTest.java]: Yes but not in TmsAgentTest: in ClusterTest.

File: management/tms-entity/tms-entity-tests/src/test/java/org/terracotta/management/entity/tms/TmsAgentTest.java (hunk 3 above)
Comment (mathieucarbou) [inline management/tms-entity/tms-entity-tests/src/test/java/org/terracotta/management/entity/tms/TmsAgentTest.java]: Yes but not in TmsAgentTest: in ClusterTest.

File: management/monitoring-service/src/main/java/org/terracotta/management/service/monitoring/MonitoringService.java [hunk 5]
```diff
@@ -17,265 +17,39 @@
...
-import java.util.logging.Logger;
-
-import static org.terracotta.management.service.monitoring.Mutation.Type.ADDITION;
-import static org.terracotta.management.service.monitoring.Mutation.Type.CHANGE;
-import static org.terracotta.management.service.monitoring.Mutation.Type.REMOVAL;
 
 /**
  * @author Mathieu Carbou
  */
 class MonitoringService {
```
Comment (anthonydahanne) [inline management/monitoring-service/src/main/java/org/terracotta/management/service/monitoring/MonitoringService.java]: there's also an interface with the same name.
I would suggest we name the interface IMonitoringService, but.. I don't really like it either...

File: management/monitoring-service-entity/src/test/java/org/terracotta/management/entity/monitoring/MonitoringServiceTest.java [hunk 6]
```diff
@@ -44,7 +44,7 @@
 @RunWith(JUnit4.class)
 public class MonitoringServiceTest {
```
Comment (anthonydahanne) [inline management/monitoring-service-entity/src/test/java/org/terracotta/management/entity/monitoring/MonitoringServiceTest.java]: there's already a test class named MonitoringServiceTest
interesting part is that neither of them "directly" test MonitoringService: one uses the provider to fetch the Monitoring Service, the other one interacts with the system through the Passthrough
Maybe we could use names that better qualify the intents of those tests; such as : 
MonitoringTreeUsingPassthroughTest and MonitoringNotificationsAndBuffersTest ?

File: management/monitoring-service-entity/src/main/java/org/terracotta/management/entity/monitoring/MonitoringService.java [hunk 7]
```diff
@@ -15,32 +15,36 @@
...
+  Serializable getValueForNode(long consumerId, String[] parents, String nodeName);
 
-  Collection<String> getChildNamesForNode(String... path);
+  <T extends Serializable> T getValueForNode(long consumerId, String[] parents, String nodeName, Class<T> type);
 
-  Collection<String> getChildNamesForNode(String[] parent, String nodeName);
+  Collection<String> getChildNamesForNode(long consumerId, String... path);
 
-  Map<String, Object> getChildValuesForNode(String... path);
+  Collection<String> getChildNamesForNode(long consumerId, String[] parent, String nodeName);
```
Comment (anthonydahanne) [inline management/monitoring-service-entity/src/main/java/org/terracotta/management/entity/monitoring/MonitoringService.java]: not used... Maybe time for deletion ?

Comment (mathieucarbou) [inline management/monitoring-service-entity/src/main/java/org/terracotta/management/entity/monitoring/MonitoringService.java]: If a remember, some methods like these are used by other projects...

**Comment by anthonydahanne** [comment]
LGTM

**Pull request merged**